docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,456 @@
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// Used to tell the runtime to generate the async state machinery for this method
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodAttribute() : Attribute();
+
+// Used to mark locals that should be hoisted to the generated async closure. Note that the runtime does not guarantee that all locals marked with this attribute will be hoisted; if it can prove that it
+// doesn't need to hoist a variable, it may avoid doing so.
+public class HoistedLocal();
+```
+
+For experimentation purposes, we recognize an attribute that can be used to force the compiler to generate the runtime async code, or to force the compiler to generate a full state machine. This attribute is not
+defined in the BCL, and exists as an escape hatch for experimentation. It may be removed when the feature ships in stable.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodGenerationAttribute(bool runtimeAsync) : Attribute();
+```
+
+## Transformation strategy
+
+As mentioned previously, we try to expose as little of this to initial binding as possible. The one major exception to this is our handling of the `RuntimeAsyncMethodAttribute`; we do not let this be applied to
+user code, and will issue an error if a user tries to do this by hand.
+
+Compiler generated async state machines and runtime generated async share some of the same building blocks. Both need to have `await`s with in `catch` and `finally` blocks rewritten to pend the exceptions,
+perform the `await` outside of the `catch`/`finally` region, and then have the exceptions restored as necessary.
+
+TODO: Go over `IAsyncEnumerable` and confirm that the initial rewrite to a `Task`-based method produces code that can then be implemented with runtime async, rather than a full compiler state machine.
+
+TODO: Clarify with the debugger team where NOPs need to be inserted for debugging/ENC scenarios.
+    We will likely need to insert AwaitYieldPoint and AwaitResumePoints for the scenarios where we emit calls to `RuntimeHelpers` async helpers, but can we avoid them for calls in runtime async form?
+
+### Example transformations
+
+Below are some examples of what IL is generated for specific examples.
+
+TODO: Include debug versions
+
+#### Await `Task`-returning method
+
+```cs
+class C
+{
+    static Task M();
+}
+
+await C.M();
+```
+
+```il
+call modreq(class [System.Runtime]System.Threading.Tasks.Task) void C::M()
+```
+
+---------------------------
+
+```cs
+class C
+{
+    Task M();
+}
+
+var c = new();
+await c.M();
+```
+
+```il
+newobj instance void C::.ctor()
+callvirt instance modreq(class [System.Runtime]System.Threading.Tasks.Task) void C::M()
+```
+
+#### Await a concrete `T` `Task<T>`-returning method
+
+```cs
+class C
+{
+    static Task<int> M();
+}
+
+await C.M();
```

Comment:
nit: Consider assigning the resulting `int` to a local.
Same comment applies to other `Task'1` scenarios